### PR TITLE
[WIP] Quantization dev main add rewrite utils make pad allow dynamic

### PR DESCRIPTION
**Pull request by AndrewZhaoLuo**
Allow a dynamic padding value but return a tensor with static shape. Currently if a dynamic padding value is used, the output tensor has a dynamic shape even if it can be statically calculated via static pad_widths.

Run: 
`python python/tvm/relay/transform/quantization/quantized_operators/conv.py`

Get:
```
tensor type `float32` has 0 dimensions, while `Tensor[(1, 2, 5, 5), float32]` has 4 dimensions
The Relay type checker is unable to show the following types match.
In particular `Tensor[(1, 2, 5, 5), float32]` does not match `float32`
The type inference pass was unable to infer a type for this expression.
This usually occurs when an operator call is under constrained in some way, check other reported errors for hints of what may of happened.
The type inference pass was unable to infer a type for this expression.
This usually occurs when an operator call is under constrained in some way, check other reported errors for hints of what may of happened.
note: run with `TVM_BACKTRACE=1` environment variable to display a backtrace.
```

### CODE REVIEW

File: python/tvm/relay/op/nn/nn.py [hunk 1]
```diff
@@ -1606,15 +1606,13 @@ def pad(data, pad_width, pad_value=0, pad_mode="constant"):
...
-    if isinstance(pad_value, Constant):
-        pad_value = pad_value.data.asnumpy().item()
     if isinstance(pad_width, Constant):
         pad_width = [list(i) for i in pad_width.data.asnumpy()]
-    if isinstance(pad_width, Expr) or (isinstance(pad_value, Expr)):
+    if not isinstance(pad_value, Expr):
+        pad_value = const(pad_value)
+    if isinstance(pad_width, Expr):
         if not isinstance(pad_width, Expr):
             pad_width = const(list(pad_width))
```
Comment (mbrookhart) [inline python/tvm/relay/op/nn/nn.py]: The second if here cannot be true, you can remove it (and the included scope).

File: python/tvm/relay/transform/quantization/quantized_operators/conv.py [hunk 2]
```diff
@@ -283,7 +283,7 @@ def example_conv_no_zp(in_channels, out_channels, img_height, img_width, groups=
     weight = relay.Var("weight")
     pad_data = relay.nn.pad(data, ((0, 0), (0, 0), (1, 1), (1, 1)), pad_value=pad_value)
     out = relay.nn.conv2d(pad_data, weight)
-    f = relay.Function([data, weight, pad_value], out)
+    f = relay.Function([data, weight, pad_value], pad_data)
```
Comment (mbrookhart) [inline python/tvm/relay/transform/quantization/quantized_operators/conv.py]: You definitely need to return out here.

File: src/relay/op/nn/pad.cc [hunk 3]
```diff
@@ -170,20 +170,19 @@ Array<te::Tensor> PadCompute(const Attrs& attrs, const Array<te::Tensor>& inputs
...
     pad_after.push_back(pad_width[i][1]);
   }
-  const auto* out_ttype = out_type.as<TensorTypeNode>();
-  return Array<te::Tensor>{topi::pad(inputs[0], pad_before, pad_after,
-                                     tvm::tir::make_const(out_ttype->dtype, param->pad_value),
-                                     "T_pad", topi::kElementWise, param->pad_mode)};
+  // const auto* out_ttype = out_type.as<TensorTypeNode>();
+  const PrimExpr& pad_value = inputs[1](Array<PrimExpr>());
+  return Array<te::Tensor>{topi::pad(inputs[0], pad_before, pad_after, pad_value, "T_pad",
+                                     topi::kElementWise, param->pad_mode)};
```
Comment (mbrookhart) [inline src/relay/op/nn/pad.cc:176]: This seems wrong to be, but I'm not exactly sure why, will look more

File: python/tvm/relay/transform/quantization/quantized_operators/conv.py (hunk 2 above)
Comment (AndrewZhaoLuo) [reply python/tvm/relay/transform/quantization/quantized_operators/conv.py]: Yep that did it.

**Comment by AndrewZhaoLuo** [comment]
Thanks Matthew for helping out for this! I'm going to make a new branch to merge this into main.

**Pull request closed without merging**